### PR TITLE
Feature/extended test suite

### PR DESCRIPTION
**Pull request by ckieschnick**
Added and extended the test suite.

Improved/fixed several API parts. There are still TODOs scattered around which should be cleared before the final merge.

### CODE REVIEW

File: src/partial17.lib/partial17/Partial.h [hunk 1]
```diff
@@ -195,13 +193,22 @@ struct Partial {
...
+
     template<class F>
     auto visitAll(F&& f) {
-        visitInitialized(std::forward<F>(f), m_data.get());
+        visitIndexTypes(f, Indices{});
     }
     template<class F>
     auto visitAll(F&& f) const {
-        visitInitialized(std::forward<F>(f), m_data.get());
+        visitIndexTypes(f, Indices{});
```
Comment (arBmind) [inline src/partial17.lib/partial17/Partial.h]: Please use `std::forward` here. Otherwise the lambdas with captures might not be able to use this.

File: src/strong17.lib/strong17/Strong.inspect.h [hunk 2]
```diff
@@ -17,10 +17,10 @@ constexpr auto hasTag(Type<Strong<Value, Tags...>>, Type<CandidateTag> = {}) {
...
 }
 
-/// tag inspection
-template<class CandidateTag, class Value, class... Tags>
-constexpr auto contains(Type<Strong<Value, Tags...>>, Type<CandidateTag> = {}) {
-    return (std::is_same_v<CandidateTag, Tags> || ...);
+/// type inspection
+template<class CandidateType, class Value, class... Tags>
+constexpr auto hasType(Type<Strong<Value, Tags...>>, Type<CandidateType> = {}) {
+    return std::is_same_v<Value, CandidateType>;
```
Comment (arBmind) [inline src/strong17.lib/strong17/Strong.inspect.h:23]: ist a `hasType` function really necessary? - we have `extractValueType` already and it's easy to compare two types.

File: src/strong17.lib/strong17/Strong.manip.test.cpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    static_assert(
+        type<AddStrongTag<DistanceTag, RemoveStrongTag<PositionTag, PositionExplicit>>> == type<DistanceExplicit>);
+    static_assert(type<ChangeStrongValue<float, PositionExplicit>> == type<Strong<float, PositionTag>>);
+    static_assert(type<ReplaceStrongTag<PositionTag, DistanceTag, PositionExplicit>> == type<DistanceExplicit>);
+}
+
+STRONG_OPAQUE(PositionOpaque, int, PositionTag);
+STRONG_OPAQUE(DistanceOpaque, int, DistanceTag);
+
+// TODO CK: Should it be possible to manipulate Opaque Strongs?
```
Comment (arBmind) [inline src/strong17.lib/strong17/Strong.manip.test.cpp]: That is not really intended. Right now the pattern is to decay the strongness with `Base` change the type and make it opaque again.

File: src/tuple17.lib/tuple17/Tuple.ostream.test.cpp [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    using T = Tuple<int, double, char, short, bool>;
+    const T t{2, 4.2, 'c', 3, false};
+    std::stringstream actual, expected;
+    actual << t;
+    expected << "[2; 4.2; c; 3; 0]";
+    EXPECT_EQ(actual.str(), expected.str());
+}
+
+TEST(Tuple, ostreamTODO_1) {
+    // TODO CK: Does not work for multiple use - ostream should work using indices
```
Comment (arBmind) [inline src/tuple17.lib/tuple17/Tuple.ostream.test.cpp:18]: 👍 - but we do not use it that way, right now. So this seems a nice to have.

File: src/tuple17.lib/tuple17/Tuple.test.cpp [hunk 5]
```diff
@@ -10,38 +10,202 @@
...
+    SingleArg s0{23};
+    SingleArg s1(s0); // Test that it won't use the argument forward constructor
 
-    X a{23};
-    X b(a); // Test that it won't use the argument forward constructor
+    EXPECT_EQ(s0, s1);
+
+    using Ambiguous = Tuple<char, int, int>;
+    auto a0 = Ambiguous{};
+    // TODO CK: Tuple is sorted, should accept only sorted list
```
Comment (arBmind) [inline src/tuple17.lib/tuple17/Tuple.test.cpp:165]: 👍 - sounds legit.

File: src/tuple17.lib/tuple17/Tuple.test.cpp [hunk 6]
```diff
@@ -10,38 +10,202 @@
...
+    auto l = Large(s);
+
+    EXPECT_EQ((l.of<char>()), 'c');
+    EXPECT_EQ((l.of<float>()), 0);
+
+    EXPECT_EQ((l.at<0>()), 23);
+    EXPECT_EQ((l.at<1>()), 0);
+    EXPECT_EQ((l.at<2>()), 'c');
+    // TODO CK: Value 23 - what is expected? I think it should prevent the creation as incompatible!
+    // ASSERT_EQ((l.at<3>()), 0);
```
Comment (arBmind) [inline src/tuple17.lib/tuple17/Tuple.test.cpp:200]: Yes this is weird but I don't want to spend time on thinking about these corner cases, until we have a use case.
Tuple should be fast and correct for a "custom" definition of input spaces. Strange behavior in unused corner cases, should not bother us.

File: src/partial17.lib/partial17/Partial.test.cpp [hunk 7]
```diff
@@ -11,16 +12,48 @@ using namespace partial17;
...
+
+    auto p1 = P{'c', 2.3f};
+    auto p2 = p1; // copy constructor
+    auto p3 = P{std::move(p2)}; // move construct
+
+    EXPECT_EQ(p1.get<0>(), 'c');
+    EXPECT_EQ(p1.get<2>(), 2.3f);
+    EXPECT_EQ(p1.countInitialized(), 2);
+    EXPECT_EQ(p1.countAll(), 3);
+    // TODO CK: Consider alignment - does this need to consider the size of the bitset
```
Comment (arBmind) [inline src/partial17.lib/partial17/Partial.test.cpp:34]: `size()` is simply the size of the heap allocated for the partial. We might want to document this.

File: src/partial17.lib/partial17/Partial.test.cpp [hunk 8]
```diff
@@ -11,16 +12,48 @@ using namespace partial17;
...
-    x = y; // copy operator
-    x = std::move(y); // move operator
-    auto z = Partial<char, int, float>{std::move(x)}; // move construct
+TEST(Partial, construction) {
+    using P = Partial<char, int, float>;
+    auto p0 = P{};
+    ASSERT_FALSE(p0.has<0>());
+    ASSERT_FALSE(p0.has<1>());
+    ASSERT_FALSE(p0.has<2>());
+    // TODO CK: Should this be possible? - crashes, would suggest a graceful value for production
```
Comment (arBmind) [inline src/partial17.lib/partial17/Partial.test.cpp:21]: This is by intention. `get<N>` is like a `cast`. You have to know that the type is there, otherwise ghosts will hunt you at night. If we throw exceptions here, the resulting program would suffer a huge performance hit.

**Review by arBmind — COMMENTED** [review]
Very happy to see a serious code cleanup here.

I answered the TODO comments and some codes with review comments.

**Comment by MaxMatti** [comment]
Code coverage shows no untested runtime functions :+1:. Unfortunately the detection for template metaprogramming and constexpr functions shows false positives for untested code, so I don't see a way to automate that.

File: src/partial17.lib/partial17/Partial.test.cpp (hunk 8 above)
Comment (ckieschnick) [reply src/partial17.lib/partial17/Partial.test.cpp:21]: Is there no way to prevent an memory access crash when trying to read uninitialized memory? Since there is the `m_bitset` to check, maybe we can return a reference to some static default for production mode while crashing with an assertion in debug mode?

File: src/partial17.lib/partial17/Partial.test.cpp (hunk 7 above)
Comment (ckieschnick) [reply src/partial17.lib/partial17/Partial.test.cpp:34]: Since `size()` it returns a binary size, I would suggest changing the naming to something in that direction. The *STL* uses `size()` for the number of elements within a container (which is currently retrievend by `countAll()` and `countInitialized()`).

File: src/partial17.lib/partial17/Partial.test.cpp (hunk 8 above)
Comment (arBmind) [reply src/partial17.lib/partial17/Partial.test.cpp:21]: Checking every time kills the optimizer.
The solution would be to rename `get` into `cast` or `as` and add a `get` that is checked. Like `operator[]` and `at` on a `std::vector`. That way we have options at the usage side.

File: src/partial17.lib/partial17/Partial.test.cpp (hunk 7 above)
Comment (arBmind) [reply src/partial17.lib/partial17/Partial.test.cpp:34]: Yes, you are correct in your observation. It's best compared to `capacity` on a `std::vector` but the semantics are different. It seems the STL has the wrong names here. So I am quite undecided what is best. We need a wording scheme for these methods, document it and use it on all value types where appropriate.
